AX: attributed string for range truncated for list items

minorninth opened the pull request:
#### 98770472e072b7648f4324af7867c294e40f8fb0
<pre>
AX: attributed string for range truncated for list items
<a href="https://bugs.webkit.org/show_bug.cgi?id=287124">https://bugs.webkit.org/show_bug.cgi?id=287124</a>
<a href="https://rdar.apple.com/144274917">rdar://144274917</a>

Reviewed by Tyler Wilcock.

The list item&apos;s text node might consist of, say, 10 characters, but
the attributed string for the whole text element includes the list
marker character, resulting in a length of 12 total. But if you fetch
the attributed string for range with a range of (0, 10) you end up
getting the list marker, but the end of the string gets truncated.
The fix is to take the prefix from the list marker into account
when returning the attributed string for a range including the beginning
of the node.

* LayoutTests/accessibility/mac/line-text-marker-range-in-list-item-expected.txt: Added.
* LayoutTests/accessibility/mac/line-text-marker-range-in-list-item.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::textContentPrefixFromListMarker const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::textContentPrefixFromListMarker const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attributedStringForTextMarkerRange const):

Canonical link: <a href="https://commits.webkit.org/290041@main">https://commits.webkit.org/290041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c37d0068cda620521e122519c0af5e105622169

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93728 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39519 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90812 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16469 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68436 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26122 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6643 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80280 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48801 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34693 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38628 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76761 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35591 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95568 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15941 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77298 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76578 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18864 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20969 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19388 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15955 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17477 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1803,6 +1803,17 @@ StringView AccessibilityObject::listMarkerTextForNodeAndPosition(Node* node, Pos
 
+String AccessibilityObject::textContentPrefixFromListMarker() const
+{
+    std::optional<SimpleRange> firstCharRange = rangeForCharacterRange({ 0, 1 });
```
Do we always only want to look at that first character/length 1 here? Are the places where there could be a bigger prefix?

minorninth replied:
So to clarify what we're doing here is getting the attributed string for range (0, 1). If this is a bulleted list with text "Hello" we'll get "• H". If this was list item 999 we'll get "999. H". We then ignore the last character and the rest is the prefix that came from the list marker.

hoffmanjoshua replied:
Got it! Do you think you could add a brief comment here explaining that?

minorninth replied:
Sure, done!

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1803,6 +1803,17 @@ StringView AccessibilityObject::listMarkerTextForNodeAndPosition(Node* node, Pos
 
+String AccessibilityObject::textContentPrefixFromListMarker() const
+{
+    std::optional<SimpleRange> firstCharRange = rangeForCharacterRange({ 0, 1 });
```
I know it can make things a bit verbose, but WebKit coding style typically recommends avoiding abbreviations (with exceptions for things like `i` in a for loop), so maybe `firstCharacterRange` instead?

minorninth replied:
Sure!

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1803,6 +1803,17 @@ StringView AccessibilityObject::listMarkerTextForNodeAndPosition(Node* node, Pos
+{
+    std::optional<SimpleRange> firstCharRange = rangeForCharacterRange({ 0, 1 });
+    if (firstCharRange) {
+        String firstCharText = stringForRange(*firstCharRange);
```
Same here — maybe `firstCharacterText`?

twilco commented on Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
```diff
@@ -315,6 +315,12 @@
+    // If the range spans the beginning of the node, account for the length of the text marker prefix, if any.
+    if (!nsRange->location) {
+        unsigned textMarkerPrefix = textContentPrefixFromListMarker().length();
+        nsRange->length += textMarkerPrefix;
```
I'm OK to leave this as-is — but I wonder if this local `textMarkerPrefix` variable provides value over directly appending `textContentPrefixFromListMarker().length()` like so:

```cpp
if (!nsRange->location)
    nsRange->length += textContentPrefixFromListMarker().length();
```

minorninth replied:
That works!

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-in-list-item.html:
```diff
@@ -0,0 +1,36 @@
+</ul>
+
+<script>
+  let output = "This tests that fetching the string for the line range corresponding to a text marker fetches the whole string, and it doesn't get truncated due to the list marker adding extra characters\n\n";
```
I think we go for 4-space indentation of JS in accessibility tests (and we do that below too, so it would be good to be consistent here)

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-in-list-item.html:
```diff
@@ -0,0 +1,36 @@
+    const marker = list.endTextMarkerForTextMarkerRange(fullRange);
+    const range = list.lineTextMarkerRangeForTextMarker(marker);
+    var rangeStr = (list.stringForTextMarkerRange(range) + "").trim();
+    output += 'stringForTextMarkerRange: ' + rangeStr + '\n';
```
Another option would be to use:

```js
output += `stringForTextMarkerRange ${rangeStr}\n`;
```

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-in-list-item.html:
```diff
@@ -0,0 +1,36 @@
+    var attrRangeStr = (list.attributedStringForTextMarkerRange(range) + "").trim();
+    // Filter out lines with AX attributes because they're not identical on all systems
+    attrRangeStr = attrRangeStr.split("\n").filter(line => line.indexOf("AX") == -1).join(" ");
+    output += 'attributedStringForTextMarkerRange: ' + attrRangeStr + '\n';
```
Another place where we could consider format strings:

```js
output += `attributedStringForTextMarkerRange: ${attrRangeStr}\n`;
```

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1998,6 +1999,13 @@ void AXIsolatedObject::setAccessibleName(const AtomString&)
+{
+    if (m_propertyMap.contains(AXProperty::TextContentPrefixFromListMarker))
+        return propertyValue<String>(AXProperty::TextContentPrefixFromListMarker);
+    return { };
```
Do we need to check if the map contains this property? Or would it be equivalent (and one less hash map lookup) to do:

```cpp
return propertyValue<String>(AXProperty::TextContentPrefixFromListMarker);
```

If this way works, we could even inline it in the header too (though the compiler would probably inline a one-liner function anyways)

minorninth replied:
You're right, that works. That suggests we can fix some other similar code that calls contains before propertyValue.

twilco reviewed (approved):
r+ with comments — thank you for fixing this!

minorninth committed:
AX: attributed string for range truncated for list items
https://bugs.webkit.org/show_bug.cgi?id=287124
rdar://144274917

Reviewed by Tyler Wilcock.

The list item's text node might consist of, say, 10 characters, but
the attributed string for the whole text element includes the list
marker character, resulting in a length of 12 total. But if you fetch
the attributed string for range with a range of (0, 10) you end up
getting the list marker, but the end of the string gets truncated.
The fix is to take the prefix from the list marker into account
when returning the attributed string for a range including the beginning
of the node.

* LayoutTests/accessibility/mac/line-text-marker-range-in-list-item-expected.txt: Added.
* LayoutTests/accessibility/mac/line-text-marker-range-in-list-item.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::textContentPrefixFromListMarker const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::textContentPrefixFromListMarker const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attributedStringForTextMarkerRange const):

Canonical link: https://commits.webkit.org/290041@main
